### ISSUE TITLE
Custom rendering, performance (questions)

### ISSUE THREAD

**Issue by teslor** [TonyGermaneri/canvas-datagrid#106]
Hi, I'm trying to use the lib in my project and have some questions.
1) What is optimal (related to performance) way to create interactive clickable elements (like buttons, checkboxes etc.) inside cells? Should I manually draw them in renderCell event and look at coordinates in click event to check the bounds? How can I change the cell style after click and manually rerender a cell?
2) I have many cell updates per second (several thousands) and the lib has good performance only in Chrome, in FF/IE it is much slower. What is the possible reason?
3) What is the best way to add many new rows to the table at once. I noticed that replacing the entire data works much faster then using multiple calls of addRow method.
4) How to make specific cells readonly?

Thanks!

**Comment by TonyGermaneri** [comment]
> 1. What is optimal (related to performance) way to create interactive clickable elements

`afterrendercell` is a good place to do it.

Here's an example of lots of drawing and updating:
https://tonygermaneri.github.io/canvas-datagrid/tutorials/sparklineDemo.html

> 1a. ...and look at coordinates in click event to check the bounds?

Correct.  You can get most of the data from `e.cell` that is pertinent to your task:

> 2. I have many cell updates per second (several thousands) and the lib has good performance only in Chrome, in FF/IE it is much slower. What is the possible reason?

The project has yet to undergo any optimization, I'm still adding features and until they are more or less done I don't want to optimize much.

That said, I don't think you need to update all the cells.  Try updating just the cells that are visible at the moment.  Check out [https://tonygermaneri.github.io/canvas-datagrid/docs/#canvasDatagrid.isCellVisible](isCellVisible) (which appears to be in the docs twice by mistake I just discovered)

> 3. What is the best way to add many new rows to the table at once.

You can manipulate the data array directly e.g.: `grid.data.push({newStuff: 'foo'});` using native array methods (or array prototypes).  Because it's impossible for the grid to tell you've done that, you need to manually call `grid.draw();` when you do it if you want to see it updated.  You can wait for the user to move their mouse over the grid causing a draw, but `grid.draw();` is very inexpensive.

As you stated you can replace the array, but believe it or not, that's a little slower because `grid.data` is a setter that tries (for better or worse #67) to corerse your data into a common format.  That's not super slow, but it's certainly slower than doing nothing.

TLDR; `grid.data.push()`, `grid.data.concat()` etc. are probably the fastest way to add records to an existing set but you need to call `grid.draw()` after doing it.

> 4. How to make specific cells readonly?

https://tonygermaneri.github.io/canvas-datagrid/docs/#canvasDatagrid#event:beforebeginedit

(FYI below sample code was written in this github editor just now and never executed, copy/paste at your own risk)

    grid.addEventListener('beforebeginedit', function (e) {
        if (/readOnlyColumn/.test(e.header.name)) {
            e.preventDefault();
        }
    });

**Comment by teslor** [comment]
Thank you for such a detailed answer, it helped me a lot.  Currently I continue exploration of the lib.
Have a good week!
